python/ray/serve/backend_state.py

```diff
@@ -30,6 +30,129 @@
+            # creation has already been started if this class was checkpointed
+            # in the STARTING state.
+            self.start()
+        elif self._state == ReplicaState.RUNNING:
+            # Fetch actor handles for all backend replicas in the system.
+            # All of these backend_replicas are guaranteed to already exist
+            # because they would not be written to a checkpoint in
+            # until they were created.
+            self._actor_handle = ray.get_actor(self._actor_name)
+        elif self._state == ReplicaState.STOPPING:
+            self.stop()
+
+    def start(self, backend_info: Optional[BackendInfo]):
+        assert self._state in {
+            ReplicaState.SHOULD_START, ReplicaState.STARTING
+        }, (f"State must be {ReplicaState.SHOULD_START} or "
+            f"{ReplicaState.STARTING}, *not* {self._state}")
+        try:
+            self._actor_handle = ray.get_actor(self._actor_name)
+        except ValueError:
+            logger.debug("Starting replica '{}' for backend '{}'.".format(
+                self._replica_tag, self._backend_tag))
+            self._actor_handle = ray.remote(backend_info.worker_class).options(
+                name=self._actor_name,
+                lifetime="detached" if self._detached else None,
+                max_restarts=-1,
+                max_task_retries=-1,
+                **backend_info.replica_config.ray_actor_options).remote(
+                    self._backend_tag, self._replica_tag,
+                    backend_info.replica_config.actor_init_args,
+                    backend_info.backend_config, self._controller_name)
+        self._startup_obj_ref = self._actor_handle.ready.remote()
+        self._state = ReplicaState.STARTING
+
+    def check_started(self):
+        if self._state == ReplicaState.RUNNING:
+            return True
+        assert self._state == ReplicaState.STARTING, (
+            f"State must be {ReplicaState.STARTING}, *not* {self._state}")
+        ready, _ = ray.wait([self._startup_obj_ref], timeout=0)
+        if len(ready) == 1:
+            self._state = ReplicaState.RUNNING
+            return True
+        return False
+
+    def set_should_stop(self, graceful_shutdown_timeout_s: Duration):
+        self._state = ReplicaState.SHOULD_STOP
+        self._graceful_shutdown_timeout_s = graceful_shutdown_timeout_s
+
+    def stop(self):
+        # We need to handle transitions from:
+        #  SHOULD_START -> SHOULD_STOP -> STOPPING
+        # This means that the replica_handle may not have been created.
+
+        assert self._state in {
+            ReplicaState.SHOULD_STOP, ReplicaState.STOPPING
+        }, (f"State must be {ReplicaState.SHOULD_STOP} or "
+            f"{ReplicaState.STOPPING}, *not* {self._state}")
+        # TODO(edoakes): this can be done without using an asyncio future.
+        @ray.remote
+        def kill_actor(actor_name, graceful_shutdown_timeout_s):
```

Comment:
Why change this to a remote task? We can do this without an asyncio future or a remote task by moving the `graceful_shutdown_timeout_s` logic inside of `check_stopped` -- if we `check_stopped` and the timeout has been exceeded, force kill it.

Reply:
SGTM!!